src/coreComponents/constitutive/unitTests/testInvariantImmiscibleFluid.cpp

```diff
@@ -0,0 +1,475 @@
+  }
+
+public:
+
+  // Helper method to get the fluid
+  InvariantImmiscibleFluid & getFluid()
+  {
+    return static_cast< InvariantImmiscibleFluid & >( *Base::getFluid( InvariantImmiscibleFluid::catalogName() ) );
+  }
+
+};
+
+template< typename MODEL_TYPE >
+class InvariantImmiscibleFluidTestFixture : public FluidModelTest< InvariantImmiscibleFluid, 3, 3 >
+{
+public:
+  using Base = FluidModelTest< InvariantImmiscibleFluid, 3, 3 >;
+  static constexpr bool USE_MASS = std::tuple_element_t< 0, MODEL_TYPE >::value;
+
+  static constexpr real64 relTol = 1.0e-10;
+  static constexpr real64 absTol = 1.0e-10;
+
+public:
+  InvariantImmiscibleFluidTestFixture() = default;
+
+  void SetUp() override
+  {
+    Base::createFluid( "InvariantImmiscibleFluid", [this]( InvariantImmiscibleFluid & fluid ){
+
+      // Set up component names
+      string_array & componentNames = fluid.getReference< string_array >( MultiFluidBase::viewKeyStruct::componentNamesString() );
+      componentNames.resize( 3 );
+      componentNames[0] = "H2O";
+      componentNames[1] = "C5H12";
+      componentNames[2] = "CH4";
+
+      // Set up phase names
+      string_array & phaseNames = fluid.getReference< string_array >( MultiFluidBase::viewKeyStruct::phaseNamesString() );
+      phaseNames.resize( 3 );
+      phaseNames[0] = "water";
+      phaseNames[1] = "oil";
+      phaseNames[2] = "gas";
+
+      // Set up component molar weights (water, oil, gas)
+      array1d< real64 > & componentMolarWeight = fluid.getReference< array1d< real64 > >( MultiFluidBase::viewKeyStruct::componentMolarWeightString() );
+      componentMolarWeight.resize( 3 );
+      componentMolarWeight[0] = 0.018;  // kg/mol
+      componentMolarWeight[1] = 0.2;    // kg/mol
+      componentMolarWeight[2] = 0.016;  // kg/mol
+
+      // Set up constant densities
+      array1d< real64 > & densities = fluid.getReference< array1d< real64 > >( "densities" );
+      densities.resize( 3 );
+      densities[0] = 1000.0;  // kg/m3
+      densities[1] = 800.0;   // kg/m3
+      densities[2] = 100.0;   // kg/m3
+
+      // Set up constant viscosities
+      array1d< real64 > & viscosities = fluid.getReference< array1d< real64 > >( "viscosities" );
+      viscosities.resize( 3 );
+      viscosities[0] = 0.001;     // Pa.s
+      viscosities[1] = 0.005;     // Pa.s
+      viscosities[2] = 0.00002;   // Pa.s
+
+      fluid.setMassFlag( USE_MASS );
+
+      // Need to allocate data for the fluid properties since we're now using KernelWrapper
+      fluid.initialize( 1, 1 );
+
+      // Initialize the fluid by computing properties with a test composition
+      stackArray1d< real64, 3 > composition( 3 );
+      composition[0] = 0.4;
+      composition[1] = 0.3;
+      composition[2] = 0.3;
+
+      arraySlice1d< real64 const, compflow::USD_COMP - 1 > compositionS( composition );
+      StackArray< real64, 3, 3, multifluid::LAYOUT_PHASE > compositionData( 1, 1, 3 );
+      auto composition_c = compositionData[0][0];
```

Comment:
```suggestion
      StackArray< real64, 2, 3, compflow::LAYOUT_COMP > compositionData( 1, 3 );
      auto composition_c = compositionData[0];
```

Reply:
Thank you for the suggestion, @dkachuma. Using slices as types in the fluid signatures is indeed non-intuitive. It may be worth a refactor in the near future to simplify the signatures, perhaps using views only.